### PR TITLE
apps/examples/testcase: Add testcase for security API

### PR DESCRIPTION
**Pull request by jylee9613**
To detect crypto hang issue of security API, add repeat testcases.

### CODE REVIEW

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 1]
```diff
@@ -59,28 +59,64 @@ static security_handle g_hnd = NULL;
...
+	security_data enc = {NULL, 0};
+
+	for (int i = 0; i < sizeof(g_aes_mode_table)/sizeof(security_aes_mode); i++) {
+		/* Check whether hang occurs during repeated encrypt function */
+		for (int j = 0; j < 10; j++) {
+			security_aes_mode mode = g_aes_mode_table[i];
+			security_aes_param param = {mode, iv, iv_len};
+
+			security_error res = crypto_aes_encryption(g_hnd, &param, UTC_CRYPTO_KEY_NAME,
+													&plain, &enc);
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:83]: When do you change a line? You don't do in line 70.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 2]
```diff
@@ -59,28 +59,64 @@ static security_handle g_hnd = NULL;
...
+
+/**
+ * @testcase         utc_crypto_aes_encryption_iv_n_p
+ * @brief            encrypt AES
+ * @scenario         encrypt AES
+ * @apicovered       crypto_aes_encryption
+ * @precondition     AES key should be set in Factory 
+ * @postcondition    none
+ */
+static void utc_crypto_aes_encryption_iv_n_p(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:98]: What does `n_p` mean?

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 3]
```diff
@@ -59,28 +59,64 @@ static security_handle g_hnd = NULL;
...
-		TC_SUCCESS_RESULT();
+	for (int i = 0; i < sizeof(g_aes_mode_table)/sizeof(security_aes_mode); i++) {	
+		/* Check whether hang occurs during repeated encrypt function */
+		for (int j = 0; j < 10; j++) {
+			security_aes_mode mode = g_aes_mode_table[i];
+			security_aes_param param = {mode, iv, iv_len};
+
+			security_error res = crypto_aes_encryption(g_hnd, &param, UTC_CRYPTO_KEY_NAME,
+													&plain, &enc);
+			TC_ASSERT_EQ("crypto_aes_encryption_p", res, SECURITY_OK);
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:117]: Not `iv_n_p` here?

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 4]
```diff
@@ -196,7 +232,7 @@ static void utc_crypto_aes_encryption_output_n(void)
 {
 	security_data enc = {NULL, 0};
 
-	unsigned char iv[] = "tamp_iv_value";
+	unsigned char iv[] = "temp_iv_value";
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:235]: This looks not a part of what you explain in the commit description.
If it's just typo fix, let's split a commit.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 5]
```diff
@@ -59,28 +59,64 @@ static security_handle g_hnd = NULL;
...
+													&plain, &enc);
+			TC_ASSERT_EQ("crypto_aes_encryption_p", res, SECURITY_OK);
+			TC_SUCCESS_RESULT();
+		}
+	}
+}
+
+/**
+ * @testcase         utc_crypto_aes_encryption_iv_n_p
+ * @brief            encrypt AES
```
Comment (kishore-sn) [inline apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:92]: The brief is same for all apis. It does not give information about what the api is doing. How about chaning it to "Negative testcase for aes iv encryption" ?  I am not sure what iv is in this case. Maybe we can elaborate if required.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 6]
```diff
@@ -56,31 +58,67 @@ static security_handle g_hnd = NULL;
...
 	security_data enc = {NULL, 0};
 
-	for (; i < sizeof(g_aes_mode_table)/sizeof(security_aes_mode); i++) {
-		security_aes_mode mode = g_aes_mode_table[i];
-		security_aes_param param = {mode, iv, iv_len};
-		security_error res = crypto_aes_encryption(g_hnd, &param, UTC_CRYPTO_KEY_NAME,
-												   &plain, &enc);
-		TC_ASSERT_EQ("crypto_aes_encryption", res, SECURITY_OK);
-		TC_SUCCESS_RESULT();
+	for (int aes_mode = 0; aes_mode < sizeof(g_aes_mode_table) / sizeof(security_aes_mode); aes_mode++) {
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:78]: 2. What is a difference between g_aes_mode_table and security_aes_mode? Why do you need to define g_aes_mode_table?
3. Could you let me know what you want from sizeof(security_aes_mode)? sizeof(int)? Could you let me know purpose of it?

You didn't reply my previous comments. Please let me get them.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 7]
```diff
@@ -59,28 +59,64 @@ static security_handle g_hnd = NULL;
...
 
-	unsigned char plain_text[] = "plain text";
-	unsigned int plain_text_len = strlen((const char*)plain_text) + 1;
+	unsigned char plain_text[] = { 0x4d, 0x79, 0x20, 0x42, 0x79, 0x74, 0x65, 0x20, 0x50, 0x72, 0x69, 0x6e, 0x74, 0x00, 0x00, 0x00 };
+	unsigned int plain_text_len = 16;
+
+	security_data plain = {plain_text, plain_text_len};
+	security_data enc = {NULL, 0};
+
+	for (int i = 0; i < sizeof(g_aes_mode_table)/sizeof(security_aes_mode); i++) {
```
[the remark this answers is not in the document]
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c]: 1. I revised it. 
2. security_aes_mode is the typedef of aes parameter, and g_aes_mode_table is actual type table to use security_aes_mode.
3. `sizeof(g_aes_mode_table) / sizeof(security_aes_mode)` is the number of elements of g_aes_mode_table, to test all types of aes mode.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 2 above)
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:100]: `iv_n` means iv is null, and `p` means pass. It is a testcase that passes when IV is null. 
aes_decrypt hang issue occurred when IV is null(using IV in secure storage), so I added this testcase.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 5 above)
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c]: I wrote the brief in detail.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 1 above)
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:85]: The existing alignment was like this, there was no intent. If I need to align, I'll change it.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 8]
```diff
@@ -209,31 +247,67 @@ static void utc_crypto_aes_encryption_output_n(void)
...
+			TC_ASSERT_EQ("crypto_aes_decryption_p", res, SECURITY_OK);
+			TC_SUCCESS_RESULT();
+		}
+	}
+}
+
+/**
+ * @testcase         utc_crypto_aes_decryption_iv_n_p
+ * @brief            decrypt AES without IV
+ * @scenario         decrypt AES without IV (IV is NULL, using IV in secure storage)
```
[the remark this answers is not in the document]
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:289]: I write in the precondition.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c [hunk 9]
```diff
@@ -56,31 +58,67 @@ static security_handle g_hnd = NULL;
...
-	int i = 0;
-	unsigned char iv[] = "temp_iv_value";
-	unsigned int iv_len = strlen((const char*)iv) + 1;
+	unsigned char iv[] = { 0x1a, 0x2b, 0x3c, 0x4d, 0x5e, 0x6f, 0x70, 0x81, 0x92, 0xa3, 0xb4, 0xc5, 0xd6, 0xe7, 0xf8, 0x00 };
+	unsigned int iv_len = 16;
 
-	unsigned char plain_text[] = "plain text";
-	unsigned int plain_text_len = strlen((const char*)plain_text) + 1;
+	unsigned char plain_text[] = { 0x4d, 0x79, 0x20, 0x42, 0x79, 0x74, 0x65, 0x20, 0x50, 0x72, 0x69, 0x6e, 0x74, 0x00, 0x00, 0x00 };
+	unsigned int plain_text_len = 16;
```
[the remark this answers is not in the document]
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:75]: It works, but I think it's better to use variable for code readability

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 6 above)
Comment (jylee9613) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:80]: I wrote above your review.

2. security_aes_mode is the typedef of aes parameter, and g_aes_mode_table is actual type table to use security_aes_mode.
3. sizeof(g_aes_mode_table) / sizeof(security_aes_mode) is the number of elements of g_aes_mode_table, to test all types of aes mode.

Comment (sunghan-chang) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:80]: 2. I can't find any difference between `g_aes_mode_table` and `security_aes_mode`. Why don't you use `security_aes_mode` directly? Why do you need to define `g_aes_mode_table`?
3. Is that a common way for enum? Don't you use a last enum type to get number of member (size)?

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 1 above)
Comment (sunghan-chang) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:87]: Anything is ok. But let's use the same way in the same file at least.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 2 above)
Comment (sunghan-chang) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:102]: `_n` at end of function name means that it's a negative test.
`_p` means a positive test.
`_n_p` means it has both.

This name looks inproper.

File: apps/examples/testcase/ta_tc/security/utc/utc_crypto.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply apps/examples/testcase/ta_tc/security/utc/utc_crypto.c:289]: How about `(If IV is not given in the crypto function, default IV which is in the secure storage is used)`?

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**